chain/chain/src/types.rs

```diff
@@ -118,7 +118,6 @@ pub struct BlockHeaderInfo {
     pub chunk_mask: Vec<bool>,
     pub total_supply: Balance,
     pub latest_protocol_version: ProtocolVersion,
-    #[cfg(feature = "protocol_feature_rectify_inflation")]
```

Comment:
How those changes affect databases and RPC layouts?

Reply:
Have we considered removing all protocol-specific fields and replacing those with single extension point, like I sketeched in this example
```
#[derive(Debug)]
struct Data {
    pub extensions: Vec<Option<Extensions>>,
}

fn set_ext(d: &mut Data, e: Extensions) {
    let index = index_of(&e);
    println!("index is {}", index);
    if index <= d.extensions.len() {
        d.extensions.resize_with(index + 1, || None);
    }
    d.extensions[index] = Some(e);
}

fn get_F2(d: &Data) -> i64 {
    match d.extensions[2] {
        Some(Extensions::F2(value)) => value,
        _ => panic!()
    }
}

#[test]
fn extensions() {
    let mut d = Data { extensions: vec!() };
    set_ext(&mut d, Extensions::F1(1));
    set_ext(&mut d, Extensions::F2(2));
    set_ext(&mut d, Extensions::F3("Hello".to_string()));
    println!("d = {:?}", d);
    println!("f2 = {:?}", get_F2(&d));
}
```

Maybe @matklad can suggest better idioms here.

Reply:
I guess there are two points here:

* how do we add fields to internal code 
* how do we add fields to schemes we use for serialization

For the first one, I think "just add a field" is probably the right solution -- it's valuable that stabilization is just removal of gating, rather than a change from a dynamic lookup in a vector to a field lookup.

For serialization, I think we might already have the desired extensibility because the top-level of things we serialize are `enums`? That's a different pattren that proposed here, it's not super clear to me which one is better, I guess that depends also on the specific format we use. For example, in JSON having an additional `data` array doesn't seem that much different from just adding additional fields to the object itself. For borsh, it might make sense, as it, in a sense, transformes borsh into a self-describing format, where you encode each object as a list of key/value pairs.


From my POV, the primary problem is not the seriazliation format per se, but rather that we use `derive(Serialize)` types throughout codebase. So, unrelated change to the internals might compile in isolation, but break compatibility in a bad way. I prefer a setup where no types used internally a `Serizlize` and instead a bit of boilerplate code is written explicitly to convert between storage format and domain objects (as outilned in https://github.com/near/nearcore/pull/4028#discussion_r587352294)

Reply:
> How those changes affect databases and RPC layouts?

There is a database migration